plugins/Disintegrator/Disintegrator.cpp

```diff
@@ -0,0 +1,240 @@
+/*
+ * Disintegrator.cpp
+ *
+ * Copyright (c) 2019 Lost Robot <r94231@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "Disintegrator.h"
+
+#include "embed.h"
+#include "lmms_math.h"
+#include "plugin_export.h"
+
+extern "C"
+{
+
+Plugin::Descriptor PLUGIN_EXPORT disintegrator_plugin_descriptor =
+{
+	STRINGIFY(PLUGIN_NAME),
+	"Disintegrator",
+	QT_TRANSLATE_NOOP("pluginBrowser", "A delay modulation effect for very aggressive digital distortion."),
+	"Lost Robot <r94231@gmail.com>",
+	0x0100,
+	Plugin::Effect,
+	new PluginPixmapLoader("logo"),
+	NULL,
+	NULL
+} ;
+
+}
+
+
+
+DisintegratorEffect::DisintegratorEffect(Model* parent, const Descriptor::SubPluginFeatures::Key* key) :
+	Effect(&disintegrator_plugin_descriptor, parent, key),
+	m_disintegratorControls(this),
+	m_needsUpdate(true)
+{
+	sampleRateChanged();
+}
+
+
+
+void DisintegratorEffect::sampleRateChanged()
+{
+	m_sampleRate = Engine::mixer()->processingSampleRate();
+	m_2PiOverSR =  F_2PI / m_sampleRate;
+	m_lp = new BasicFilters<2>(Engine::mixer()->processingSampleRate());
+	m_hp = new BasicFilters<2>(Engine::mixer()->processingSampleRate());
+	m_lp->setFilterType(0);
+	m_hp->setFilterType(1);
+	m_needsUpdate = true;
+
+	m_bufferSize = m_sampleRate * 0.01f + 1.f;
+	for (int i = 0; i < 2; ++i)
+	{
+		m_inBuf[i].resize(m_bufferSize);
+	}
+}
+
+
+
+bool DisintegratorEffect::processAudioBuffer(sampleFrame* buf, const fpp_t frames)
+{
+	if (!isEnabled() || !isRunning ())
+	{
+		return false;
+	}
+
+	double outSum = 0.0;
+	const float d = dryLevel();
+	const float w = wetLevel();
+	
+	const ValueBuffer * amountBuf = m_disintegratorControls.m_amountModel.valueBuffer();
+	const ValueBuffer * typeBuf = m_disintegratorControls.m_typeModel.valueBuffer();
+	const ValueBuffer * freqBuf = m_disintegratorControls.m_lowCutModel.valueBuffer();
+
+	// Update filters
+	if(m_needsUpdate || m_disintegratorControls.m_highCutModel.isValueChanged())
+	{
+		m_lp->calcFilterCoeffs(m_disintegratorControls.m_highCutModel.value(), 0.5);
+		
+	}
+	if(m_needsUpdate || m_disintegratorControls.m_lowCutModel.isValueChanged())
+	{
+		m_hp->calcFilterCoeffs(m_disintegratorControls.m_lowCutModel.value(), 0.5);
```

Comment:
The updating of the filter coefficients should be done per frame using the value buffer, not once per buffer.

The current implementation is block size dependant, so the audio will be different depending on the block size. This is not desirable, and the effect should remain independent. 

One of the drives for sample exactness was to remove the buzzing and zipping noises, these are caused by sudden stepped changes in parameters on a per-block basis, and are removed by smoothly updating the parameters on a per-frame basis.